src/Sylius/Bundle/CoreBundle/Doctrine/ORM/ProductAssociationRepository.php

```diff
@@ -22,15 +22,22 @@ class ProductAssociationRepository extends EntityRepository implements ProductAs
 {
     public function findWithProductsWithinChannel($associationId, ChannelInterface $channel): ProductAssociationInterface
     {
-        return $this->createQueryBuilder('o')
+        $productAssociation = $this->createQueryBuilder('o')
             ->addSelect('associatedProduct')
-            ->innerJoin('o.associatedProducts', 'associatedProduct', 'WITH', 'associatedProduct.enabled = true')
+            ->leftJoin('o.associatedProducts', 'associatedProduct', 'WITH', 'associatedProduct.enabled = true')
             ->innerJoin('associatedProduct.channels', 'channel', 'WITH', 'channel = :channel')
             ->andWhere('o.id = :associationId')
             ->setParameter('associationId', $associationId)
             ->setParameter('channel', $channel)
             ->getQuery()
             ->getOneOrNullResult()
         ;
+
+        if (null === $productAssociation) {
+            $productAssociation = $this->find($associationId);
+            $productAssociation->clearAssociatedProducts();
+        }
```

Comment:
It's not a perfect solution but I can't achieve it with a single query. Left join for `associatedProduct.channels` results that filtering out products outside a given channel doesn't work (even if I put `channel` to the select statement).